File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/existentialWrapperViaObjCInterop/existentialWrapperViaObjCInterop.kt
// KIND: STANDALONE
// WITH_PLATFORM_LIBS
// MODULE: ExistentialWrapperViaObjCInterop
// FILE: main.kt
@file:OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)

import kotlinx.cinterop.convert
import kotlinx.cinterop.objcPtr
import platform.Foundation.NSMutableArray
import platform.Foundation.NSStringFromClass
import platform.objc.object_getClass

// todo: remove me after KT-87457
interface Anchor {
    fun ping(): Int
}

open class Exported

// Not exported and without an exported ancestor: bridging it to Objective-C makes the runtime build a
// `_KotlinExistential` wrapper.
private class Hidden

// Not exported, but with an exported ancestor, so it resolves to `Exported`'s bound wrapper class instead.
private class HiddenChild : Exported()

// Handing a Kotlin object to Objective-C goes through `Kotlin_Interop_refToObjC`
private fun roundTripThroughObjC(value: Any): Any? =
    NSMutableArray().apply { addObject(value) }.objectAtIndex(0.convert())

private val Any.objCClassName: String
    get() = NSStringFromClass(object_getClass(this)!!)

fun exportedRoundTripsThroughObjC(): Boolean = Exported().let { roundTripThroughObjC(it) === it }

fun hiddenRoundTripsThroughObjC(): Boolean = Hidden().let { roundTripThroughObjC(it) === it }

fun hiddenChildRoundTripsThroughObjC(): Boolean = HiddenChild().let { roundTripThroughObjC(it) === it }

fun unitRoundTripsThroughObjC(): Boolean = roundTripThroughObjC(Unit) === Unit

// Which of the two wrapper resolutions the interop layer used is visible in the wrapper's Objective-C class.
// `classWrapperFor()` hands a non-exported class the bound wrapper class of its closest exported ancestor;
// `protocolWrapperFor()` would hand it an uncached existential instead. So this equality is exactly the
// `conformsTo:` predicate in `+_createRetainedWrapperForKotlinObject:` being always-true rather than `nil`.
fun hiddenChildSharesWrapperClassWithExported(): Boolean =
    HiddenChild().objCClassName == Exported().objCClassName

// The negative control: without it the assertion above would also hold if everything resolved to `Exported`.
// `Hidden` has no exported ancestor, so it is the one case that does legitimately get an existential.
fun hiddenUsesDistinctWrapperClass(): Boolean = Hidden().objCClassName != Exported().objCClassName

// `objcPtr()` is `Kotlin_Interop_refToObjC`, so these observe that a Kotlin object keeps a single wrapper
// across repeated conversions rather than allocating a fresh one each time.
fun existentialWrapperIsCached(): Boolean = Hidden().let { it.objcPtr() == it.objcPtr() }
fun boundWrapperIsCached(): Boolean = HiddenChild().let { it.objcPtr() == it.objcPtr() }
fun exportedWrapperIsCached(): Boolean = Exported().let { it.objcPtr() == it.objcPtr() }

// Entry points for the Swift side, where the argument already carries a bound bridge wrapper of its own.
fun roundTripExportedThroughObjC(value: Exported): Exported = roundTripThroughObjC(value) as Exported

fun makeExported(): Exported = Exported()

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/existentialWrapperViaObjCInterop/existentialWrapperViaObjCInterop.swift
import ExistentialWrapperViaObjCInterop
import Testing
import ObjectiveC
import Foundation

/// Bridging a Kotlin object to Objective-C has to produce a wrapper that carries the object itself: storing
/// the value in an `id`-typed Objective-C container and reading it back must yield the same Kotlin object.
@Test
func testKotlinObjectSurvivesObjCRoundTrip() throws {
    try #require(exportedRoundTripsThroughObjC())
    try #require(hiddenRoundTripsThroughObjC())
    try #require(hiddenChildRoundTripsThroughObjC())
    try #require(unitRoundTripsThroughObjC())
}

/// Crossing into Objective-C resolves the wrapper class through `classWrapperFor()`, which follows the Kotlin
/// class hierarchy, rather than through `protocolWrapperFor()`, which would yield an existential.
@Test
func testWrapperClassFollowsKotlinHierarchy() throws {
    try #require(hiddenChildSharesWrapperClassWithExported())
    try #require(hiddenUsesDistinctWrapperClass())
}

/// A Kotlin object keeps one wrapper across repeated conversions to Objective-C.
@Test
func testWrapperIsReusedAcrossConversions() throws {
    try #require(existentialWrapperIsCached())
    try #require(boundWrapperIsCached())
    try #require(exportedWrapperIsCached())
}

/// A Swift-created instance of an exported class already carries its bound bridge wrapper, so crossing into
/// Objective-C has to resolve back to that very instance instead of allocating a second wrapper.
@Test
func testExportedInstanceSurvivesObjCRoundTrip() throws {
    let swiftCreated = Exported()
    try #require(roundTripExportedThroughObjC(value: swiftCreated) === swiftCreated)

    let kotlinCreated = makeExported()
    try #require(roundTripExportedThroughObjC(value: kotlinCreated) === kotlinCreated)
}

/// Here the Objective-C class associated with the Kotlin type is the Swift subclass itself, while the
/// resolved wrapper class is `Exported`'s bound bridge. Neither descends from the other, which is exactly
/// what a receiver-constrained wrapper lookup cannot express.
@Test
func testSwiftSubclassSurvivesObjCRoundTrip() throws {
    class SwiftSubclass: Exported {}

    let subclassInstance = SwiftSubclass()
    let returned = roundTripExportedThroughObjC(value: subclassInstance)
    try #require(returned === subclassInstance)
    try #require(returned is SwiftSubclass)
}

@Test
func testRuntimeSupportIsLinked() throws {
    let selector = NSSelectorFromString("_Kotlin_SwiftExport_wrapIntoExistential:")
    try #require(class_getClassMethod(NSObject.self, selector) != nil)
}
